File: ios/Classes/Renderer/LineRenderer.swift
import Foundation

final class LineRenderer: ChartLayerRenderer {
  func render(frame: ChartFrame) {
    guard frame.chartType == "line", frame.viewport.hasVisibleRange() else { return }

    let context = frame.context
    let chartRect = frame.mainChartRect

    context.setStrokeColor(frame.theme.lineColor)
    context.setLineWidth(2)
    context.beginPath()

    for index in frame.viewport.visibleStartIndex...frame.viewport.visibleEndIndex {
      let x = frame.viewport.xCenter(for: index, plotLeft: chartRect.minX)
      let y = frame.viewport.priceToY(
        frame.candleStore.close(at: index),
        top: chartRect.minY,
        height: chartRect.height
      )
      if index == frame.viewport.visibleStartIndex {
        context.move(to: CGPoint(x: x, y: y))
      } else {
        context.addLine(to: CGPoint(x: x, y: y))
      }
    }

    context.strokePath()
  }
}

File: android/src/main/kotlin/com/tradechart/plugin/renderer/LineRenderer.kt
package com.tradechart.plugin.renderer

import android.graphics.Path
import com.tradechart.plugin.engine.ChartFrame

class LineRenderer : ChartLayerRenderer {
    override fun render(frame: ChartFrame) {
        if (frame.chartType != "line" || !frame.viewport.hasVisibleRange()) {
            return
        }

        val chartRect = frame.mainChartRect
        val path = Path()
        for (index in frame.viewport.visibleStartIndex..frame.viewport.visibleEndIndex) {
            val centerX = frame.viewport.xCenterForIndex(index, chartRect.left)
            val y = frame.viewport.priceToY(
                frame.candleStore.closeAt(index),
                chartRect.top,
                chartRect.height(),
            )
            if (index == frame.viewport.visibleStartIndex) {
                path.moveTo(centerX, y)
            } else {
                path.lineTo(centerX, y)
            }
        }
        frame.canvas.drawPath(path, frame.theme.linePaint)
    }
}
